hpx/parallel/executors/parallel_executor.hpp

```diff
@@ -162,53 +163,44 @@ namespace hpx { namespace parallel { namespace execution
     protected:
         /// \cond NOINTERNAL
         template <typename Result, typename F, typename Iter, typename ... Ts>
-        hpx::future<void> spawn(std::vector<hpx::future<Result> >& results,
-            std::size_t base, std::size_t size, std::size_t num_tasks,
-            F const& func, Iter it, Ts const&... ts) const
+        void spawn(std::vector<hpx::future<Result> >& results,
+            lcos::local::latch& l, std::size_t base, std::size_t size,
+            std::size_t num_tasks, F const& func, Iter it, Ts const&... ts) const
         {
             if (size > num_tasks)
             {
                 // spawn hierarchical tasks
                 std::size_t chunk_size = (size + num_spread_) / num_spread_ - 1;
                 chunk_size = (std::max)(chunk_size, num_tasks);
 
-                std::vector<hpx::future<void> > tasks;
-                tasks.reserve(num_spread_);
-
-                hpx::future<void> (parallel_policy_executor::*spawn_func)(
-                        std::vector<hpx::future<Result> >&, std::size_t,
-                        std::size_t, std::size_t, F const&, Iter, Ts const&...
+                void (parallel_policy_executor::*spawn_func)(
+                        std::vector<hpx::future<Result> >&,
+                        lcos::local::latch& l, std::size_t, std::size_t,
+                        std::size_t, F const&, Iter, Ts const&...
                     ) const = &parallel_policy_executor::spawn;
 
-                while (size != 0)
+                while (size > chunk_size)
                 {
-                    std::size_t curr_chunk_size = (std::min)(chunk_size, size);
-
-                    hpx::future<void> f = async_execute(
-                        spawn_func, this, std::ref(results), base,
-                        curr_chunk_size, num_tasks, std::ref(func), it,
+                    async_execute(
```

Comment:
Any reason why you didn't use `post` instead of `async_execute` here? This would save another allocation of the shared state.

Reply:
No reason, will change.